compiler/rustc_expand/src/proc_macro.rs

```diff
@@ -124,36 +137,40 @@ impl MultiItemModifier for DeriveProcMacro {
         // altogether. See #73345.
         crate::base::ann_pretty_printing_compatibility_hack(&item, &ecx.sess);
         let input = item.to_tokens();
-        let stream = {
-            let _timer =
-                ecx.sess.prof.generic_activity_with_arg_recorder("expand_proc_macro", |recorder| {
-                    recorder.record_arg_with_span(
-                        ecx.sess.source_map(),
-                        ecx.expansion_descr(),
-                        span,
-                    );
-                });
-            let proc_macro_backtrace = ecx.ecfg.proc_macro_backtrace;
-            let strategy = exec_strategy(ecx);
-            let server = proc_macro_server::Rustc::new(ecx);
-            match self.client.run(&strategy, server, input, proc_macro_backtrace) {
-                Ok(stream) => stream,
-                Err(e) => {
-                    ecx.dcx().emit_err({
-                        errors::ProcMacroDerivePanicked {
-                            span,
-                            message: e.as_str().map(|message| {
-                                errors::ProcMacroDerivePanickedHelp { message: message.into() }
-                            }),
-                        }
-                    });
-                    return ExpandResult::Ready(vec![]);
-                }
+        let res = ty::tls::with(|tcx| {
+            // FIXME(pr-time): without flattened some (weird) tests fail, but no idea if it's correct/enough
+            let input = tcx.arena.alloc(input.flattened()) as &TokenStream;
+            let invoc_id = ecx.current_expansion.id;
+            let invoc_expn_data = invoc_id.expn_data();
+
+            assert_eq!(invoc_expn_data.call_site, span);
+
+            // FIXME(pr-time): Is this the correct way to check for incremental compilation (as
+            // well as for `cache_proc_macros`)?
+            if tcx.sess.opts.incremental.is_some() && tcx.sess.opts.unstable_opts.cache_proc_macros
+            {
+                // FIXME(pr-time): Just using the crate hash to notice when the proc-macro code has
+                // changed. How to *correctly* depend on exactly the macro definition?
+                // I.e., depending on the crate hash is just a HACK, and ideally the dependency would be
+                // more narrow.
```

Comment:
I don't think there's any more precise way.
Proc macro is an executable code, it can run arbitrary functions from the macro crate itself or other crates, we don't have any system for incremental tracking of runtime dependencies, so we can only assume that if anything in the crate changes it can change the macro logic as well.